receiver/clickhousesystemtablesreceiver/receiver.go

```diff
@@ -0,0 +1,205 @@
+package clickhousesystemtablesreceiver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/receiver/receiverhelper"
+	"go.uber.org/zap"
+)
+
+type systemTablesReceiver struct {
+	// next scrape will include events in query_log table with
+	// nextScrapeIntervalStartTs  <= event_timestamp < (nextScrapeIntervalStartTs + scrapeIntervalSeconds)
+	scrapeIntervalSeconds uint32
+
+	// next scrape will happen only after timestamp at server is at least
+	// (nextScrapeIntervalStartTs + scrapeIntervalSeconds) + scrapeDelaySeconds
+	scrapeDelaySeconds uint32
+
+	clickhouse   clickhouseQuerrier
+	nextConsumer consumer.Logs
+
+	logger  *zap.Logger
+	obsrecv *receiverhelper.ObsReport
+
+	nextScrapeIntervalStartTs uint32
+
+	requestShutdown    context.CancelFunc
+	shutdownCompleteWg sync.WaitGroup
+}
+
+func (r *systemTablesReceiver) Start(
+	ctx context.Context, host component.Host,
+) error {
+	err := r.init(ctx)
+	if err != nil {
+		return err
+	}
+
+	receiverCtx, cancelReceiverCtx := context.WithCancel(context.Background())
+	r.requestShutdown = cancelReceiverCtx
+
+	go func() {
+		if err := r.run(receiverCtx); err != nil {
+			host.ReportFatalError(err)
+		}
+	}()
+
+	return nil
+}
+
+func (r *systemTablesReceiver) init(ctx context.Context) error {
+	// The receiver starts scraping query_log entries that come at or after
+	// the timestamp at clickhouse server when the receiver is started.
+	serverTsNow, err := r.clickhouse.unixTsNow(ctx)
+	if err != nil {
+		return fmt.Errorf("couldn't determine ts at clickhouse at receiver startup: %w", err)
+	}
+	r.nextScrapeIntervalStartTs = serverTsNow
+
+	return nil
+}
+
+func (r *systemTablesReceiver) Shutdown(context.Context) error {
+	r.requestShutdown()
+	r.shutdownCompleteWg.Wait()
+	return nil
+}
+
+func (r *systemTablesReceiver) run(ctx context.Context) error {
+	r.shutdownCompleteWg.Add(1)
+	defer r.shutdownCompleteWg.Done()
+
+	minTsBeforeNextScrapeAttempt := time.Now().Unix()
+
+	for {
+		// time to shut down?
+		if ctx.Err() != nil {
```

Comment:
Is this needed? The passed context is background context and only possible error is context cancelled which is expected during the shutdown and we don't want to return error here?